Linux/O_STM/DOXY/man/man3/TX.3
.TH "TX" 3 "Sun Apr 1 2018" "C++ Software transactional Memory" \" -*- nroff -*-
.ad l
.nh
.SH NAME
TX \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <TX\&.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "void \fB_decrease_tx_nesting\fP ()"
.br
.RI "\fIRemove \fBTX\fP nesting level by one\&. \fP"
.ti -1c
.RI "void \fB_increase_tx_nesting\fP ()"
.br
.RI "\fIAdd \fBTX\fP nesting level by one\&. \fP"
.ti -1c
.RI "void \fB_print_all_tx\fP ()"
.br
.ti -1c
.RI "void \fB_register\fP (std::shared_ptr< \fBOSTM\fP > object)"
.br
.RI "\fIRegister \fBOSTM\fP pointer into STM library\&. \fP"
.ti -1c
.RI "bool \fBcommit\fP ()"
.br
.RI "\fICommit transactional changes\&. \fP"
.ti -1c
.RI "int \fBgetTest_counter\fP ()"
.br
.RI "\fIgetTest_counter TESTING ONLY!!! returning the value of the test_counter stored, number of rollbacks \fP"
.ti -1c
.RI "std::shared_ptr< \fBOSTM\fP > \fBload\fP (std::shared_ptr< \fBOSTM\fP > object)"
.br
.RI "\fIload std::shared_ptr<OSTM>, returning an std::shared_ptr<OSTM> copy of the original pointer, to work with during transaction life time \fP"
.ti -1c
.RI "void \fBostm_exit\fP ()"
.br
.RI "\fIDelete all map entries associated with the main process\&. \fP"
.ti -1c
.RI "void \fBstore\fP (std::shared_ptr< \fBOSTM\fP > object)"
.br
.RI "\fIStore transactional changes\&. \fP"
.ti -1c
.RI "\fBTX\fP (std::thread::id id)"
.br
.RI "\fIConstructor\&. \fP"
.ti -1c
.RI "\fBTX\fP (const \fBTX\fP &orig)"
.br
.RI "\fIDefault copy constructor\&. \fP"
.ti -1c
.RI "\fB~TX\fP ()"
.br
.RI "\fIDe-constructor\&. \fP"
.in -1c
.SS "Static Public Attributes"

.in +1c
.ti -1c
.RI "static int \fBtest_counter\fP = 0"
.br
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "const std::thread::id \fB_get_tx_number\fP () const "
.br
.RI "\fI_get_tx_number returning the transaction uniqe identifier \fP"
.ti -1c
.RI "void \fB_release_object_lock\fP ()"
.br
.RI "\fI_release_object_lock void, is get called from commit function, with the purpose to release the locks on all the objects participating in the transaction \fP"
.ti -1c
.RI "std::map< int, int > \fBget_thread_Map\fP ()"
.br
.RI "\fIget_thread_Map returning and map to insert to the process_map_collection as an inner value \fP"
.ti -1c
.RI "void \fBth_exit\fP ()"
.br
.RI "\fIClean up all associated values by the thread delete from working_Map_collection, it is an automated function\&. \fP"
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "int \fB_tx_nesting_level\fP"
.br
.ti -1c
.RI "std::thread::id \fBtransaction_Number\fP"
.br
.RI "\fIReturning the transaction number\&. \fP"
.ti -1c
.RI "std::map< int, std::shared_ptr< \fBOSTM\fP > > \fBworking_Map_collection\fP"
.br
.RI "\fIMAP Collection to store OSTM* parent based pointers to make invisible changes during isolated transaction\&. \fP"
.in -1c
.SS "Static Private Attributes"

.in +1c
.ti -1c
.RI "static std::map< int, std::shared_ptr< \fBOSTM\fP > > \fBmain_Process_Map_collection\fP"
.br
.RI "\fISTATIC GLOBAL MAP Collection to store OSTM* parent based pointers to control/lock and compare objects version number within transactions\&. \fP"
.ti -1c
.RI "static std::map< int, std::map< int, int > > \fBprocess_map_collection\fP"
.br
.RI "\fISTATIC GLOBAL MAP Collection to store all process associated keys to find when deleting transactions\&. \fP"
.ti -1c
.RI "static std::mutex \fBregister_Lock\fP"
.br
.in -1c
.SS "Friends"

.in +1c
.ti -1c
.RI "class \fBTM\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 26 of file TX\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "TX::TX (std::thread::id id)"

.PP
Constructor\&. 
.PP
\fBParameters:\fP
.RS 4
\fItransaction_Number\fP int, to store associated thread 
.br
\fI_tx_nesting_level\fP int, to store and indicate nesting level of transactions within transaction 
.RE
.PP

.PP
Definition at line 31 of file TX\&.cpp\&.
.PP
References _tx_nesting_level, and transaction_Number\&.
.PP
.nf
31                      {
32     this->transaction_Number = id;
33     this->_tx_nesting_level = 0;
34 }
.fi
.SS "TX::~TX ()"

.PP
De-constructor\&. 
.PP
Definition at line 38 of file TX\&.cpp\&.
.PP
.nf
38         {
39    
40 }
.fi
.SS "TX::TX (const \fBTX\fP & orig)"

.PP
Default copy constructor\&. 
.PP
Definition at line 44 of file TX\&.cpp\&.
.PP
.nf
44                      {
45 
46 }
.fi
.SH "Member Function Documentation"
.PP 
.SS "void TX::_decrease_tx_nesting ()"

.PP
Remove \fBTX\fP nesting level by one\&. _decrease_tx_nesting decrease the value stored in _tx_nesting_level by one, when outer transactions commiting
.PP
\fBParameters:\fP
.RS 4
\fI_tx_nesting_level\fP int 
.RE
.PP

.PP
Definition at line 316 of file TX\&.cpp\&.
.PP
References _tx_nesting_level\&.
.PP
Referenced by commit()\&.
.PP
.nf
316                               {
317    // std::cout << "[this->_tx_nesting_level] = " << this->_tx_nesting_level << std::endl;
318     this->_tx_nesting_level -= 1;
319 ;
320 }
.fi
.SS "const std::thread::id TX::_get_tx_number () const\fC [private]\fP"

.PP
_get_tx_number returning the transaction uniqe identifier _get_tx_number std::thread::id, returning the thread id that has assigned the given transaction
.PP
\fBParameters:\fP
.RS 4
\fItransaction_Number\fP int 
.RE
.PP

.PP
Definition at line 331 of file TX\&.cpp\&.
.PP
References transaction_Number\&.
.PP
.nf
331                                            {
332     return transaction_Number;
333 }
.fi
.SS "void TX::_increase_tx_nesting ()"

.PP
Add \fBTX\fP nesting level by one\&. _increase_tx_nesting increase the value stored in _tx_nesting_level by one, indicate that the transaction nested
.PP
\fBParameters:\fP
.RS 4
\fI_tx_nesting_level\fP int 
.RE
.PP

.PP
Definition at line 307 of file TX\&.cpp\&.
.PP
References _tx_nesting_level\&.
.PP
.nf
307                               {
308       
309     this->_tx_nesting_level += 1;
310     // std::cout << "[this->_tx_nesting_level] = " << this->_tx_nesting_level << std::endl;
311 }
.fi
.SS "void TX::_print_all_tx ()"
ONLY FOR TESTING CHECK THE MAP AFTER THREAD EXIT AND ALL SHOULD BE DELETED!!!!!!! 
.PP
Definition at line 346 of file TX\&.cpp\&.
.PP
References process_map_collection, and working_Map_collection\&.
.PP
.nf
346                        {
347 
348     std::cout << "[PRINTALLTHREAD]" << std::endl;
349     std::map< int, std::shared_ptr<OSTM> >::iterator it;
350     /*
351      * All registered thread id in the TX global 
352      */
353     int ppid = getpid();
354     std::map<int, std::map< int, int >>::iterator process_map_collection_Iterator = TX::process_map_collection\&.find(ppid);
355     if (process_map_collection_Iterator != TX::process_map_collection\&.end()) {
356 
357         for (auto current = process_map_collection_Iterator->second\&.begin(); current != process_map_collection_Iterator->second\&.end(); ++current) {
358             it = working_Map_collection\&.find(current->first);
359             if(it != working_Map_collection\&.end()){
360                 std::cout << "[Unique number ] : " <<it->second->Get_Unique_ID() << std::endl;
361             }
362 
363             
364         }
365      
366     }
367 }
.fi
.SS "void TX::_register (std::shared_ptr< \fBOSTM\fP > object)"

.PP
Register \fBOSTM\fP pointer into STM library\&. register void, receives an std::shared_ptr<OSTM> that point to the original memory space to protect from reca conditions
.PP
\fBParameters:\fP
.RS 4
\fIworking_Map_collection\fP std::map, store all the std::shared_ptr<OSTM> pointer in the transaction 
.br
\fImain_Process_Map_collection\fP std::map, store all std::shared_ptr<OSTM> from all transaction, used to lock and compare the objects 
.br
\fIprocess_map_collection\fP std::map, store all std::shared_ptr<OSTM> unique ID from all transaction, used to delete all pointers used by the main process, from all transaction before the program exit\&. 
.br
\fIstd::lock_guard\fP use register_Lock(mutex) shared lock between all transaction 
.br
\fIppid\fP int, store main process number 
.RE
.PP

.PP
Definition at line 104 of file TX\&.cpp\&.
.PP
References get_thread_Map(), main_Process_Map_collection, process_map_collection, register_Lock, and working_Map_collection\&.
.PP
.nf
104                                              {
105     /*
106      * MUST USE SHARED LOCK TO PROTECT SHARED GLOBAL MAP/COLLECTION 
107      */
108     std::lock_guard<std::mutex> guard(TX::register_Lock);
109     
110     /*
111      * Check for null pointer !
112      * Null pointer can cause segmentation fault!!!
113      */
114     if(object == nullptr){
115         throw std::runtime_error(std::string("[RUNTIME ERROR : NULL POINTER IN REGISTER FUNCTION]") );
116     }
117     
118     int ppid = getpid();
119     std::map<int, std::map< int, int >>::iterator process_map_collection_Iterator = TX::process_map_collection\&.find(ppid);
120     if (process_map_collection_Iterator == TX::process_map_collection\&.end()) {
121         /*
122          * Register main process/application to the global map
123          */
124         std::map< int, int >map =  get_thread_Map();
125         TX::process_map_collection\&.insert({ppid, map});
126         /*
127          * Get the map if registered first time
128          */
129         process_map_collection_Iterator = TX::process_map_collection\&.find(ppid);
130     }
131     std::map<int, std::shared_ptr<OSTM>>::iterator main_Process_Map_collection_Iterator = TX::main_Process_Map_collection\&.find(object->Get_Unique_ID());
132     if (main_Process_Map_collection_Iterator == TX::main_Process_Map_collection\&.end()) {
133         /*
134          * Insert to the GLOBAL MAP 
135          */
136         TX::main_Process_Map_collection\&.insert({object->Get_Unique_ID(), object});
137         /*
138          * Insert to the GLOBAL MAP as a helper to clean up at end of main process 
139          */
140         process_map_collection_Iterator->second\&.insert({object->Get_Unique_ID(), 1});
141     } 
142 
143 
144     std::map< int, std::shared_ptr<OSTM> >::iterator working_Map_collection_Object_Shared_Pointer_Iterator = working_Map_collection\&.find(object->Get_Unique_ID());
145     if (working_Map_collection_Object_Shared_Pointer_Iterator == working_Map_collection\&.end()) {
146 
147         working_Map_collection\&.insert({object->Get_Unique_ID(), object->getBaseCopy(object)});
148     }
149 
150 }
.fi
.SS "void TX::_release_object_lock ()\fC [private]\fP"

.PP
_release_object_lock void, is get called from commit function, with the purpose to release the locks on all the objects participating in the transaction Release the locks in objects with transaction associated collection
.PP
\fBParameters:\fP
.RS 4
\fIworking_Map_collection\fP std::map, store all the std::shared_ptr<OSTM> pointer in the transaction 
.br
\fImain_Process_Map_collection\fP std::map, store all std::shared_ptr<OSTM> from all transaction, used to release the lock on object 
.RE
.PP

.PP
Definition at line 286 of file TX\&.cpp\&.
.PP
References main_Process_Map_collection, and working_Map_collection\&.
.PP
Referenced by commit()\&.
.PP
.nf
286                              {
287     
288     std::map< int, std::shared_ptr<OSTM> >::iterator working_Map_collection_Object_Shared_Pointer_Iterator;
289     std::map<int, std::shared_ptr<OSTM>>::iterator main_Process_Map_collection_Iterator;
290     for (working_Map_collection_Object_Shared_Pointer_Iterator = working_Map_collection\&.begin(); working_Map_collection_Object_Shared_Pointer_Iterator != working_Map_collection\&.end(); working_Map_collection_Object_Shared_Pointer_Iterator++) {
291 
292             main_Process_Map_collection_Iterator = TX::main_Process_Map_collection\&.find((working_Map_collection_Object_Shared_Pointer_Iterator->second)->Get_Unique_ID());
293             if (main_Process_Map_collection_Iterator != TX::main_Process_Map_collection\&.end()) {
294                 /*
295                  * Release object lock
296                  */
297                 (main_Process_Map_collection_Iterator)->second->unlock_Mutex();
298                 
299             } 
300         }
301 }
.fi
.SS "bool TX::commit ()"

.PP
Commit transactional changes\&. commit bool, returns boolean value TRUE/FALSE depends on the action taken within the function
.PP
\fBParameters:\fP
.RS 4
\fIworking_Map_collection\fP std::map, store all the std::shared_ptr<OSTM> pointer in the transaction 
.br
\fImain_Process_Map_collection\fP std::map, store all std::shared_ptr<OSTM> from all transaction, used to lock and compare the objects 
.br
\fIcan_Commit\fP bool, helps to make decision that the transaction can commit or rollback 
.RE
.PP

.PP
Definition at line 202 of file TX\&.cpp\&.
.PP
References _decrease_tx_nesting(), _release_object_lock(), _tx_nesting_level, main_Process_Map_collection, test_counter, th_exit(), and working_Map_collection\&.
.PP
.nf
202                 {
203 
204     bool can_Commit = true;
205  
206     /*
207      * Dealing with nested transactions first 
208      */
209     if (this->_tx_nesting_level > 0) {
210         _decrease_tx_nesting();
211         return true;
212     } 
213     
214     std::map< int, std::shared_ptr<OSTM> >::iterator working_Map_collection_Object_Shared_Pointer_Iterator;
215 
216     std::map<int, std::shared_ptr<OSTM>>::iterator main_Process_Map_collection_Iterator;
217     for (working_Map_collection_Object_Shared_Pointer_Iterator = working_Map_collection\&.begin(); working_Map_collection_Object_Shared_Pointer_Iterator != working_Map_collection\&.end(); working_Map_collection_Object_Shared_Pointer_Iterator++) {
218 
219             main_Process_Map_collection_Iterator = TX::main_Process_Map_collection\&.find(working_Map_collection_Object_Shared_Pointer_Iterator->second->Get_Unique_ID());
220             /*
221              * Throws runtime error if object can not find
222              */
223             if(main_Process_Map_collection_Iterator == TX::main_Process_Map_collection\&.end())
224             {
225                 throw std::runtime_error(std::string("[RUNTIME ERROR : CAN'T FIND OBJECT COMMIT FUNCTION]"));
226             }
227 
228         /*
229          * Busy wait WHILE object locked by other thread
230          */
231         while(!(main_Process_Map_collection_Iterator->second)->is_Locked());
232 
233         if (main_Process_Map_collection_Iterator->second->Get_Version() > working_Map_collection_Object_Shared_Pointer_Iterator->second->Get_Version()) {
234 
235             working_Map_collection_Object_Shared_Pointer_Iterator->second->Set_Can_Commit(false);
236             can_Commit = false;
237             break;
238         } else {
239 
240             working_Map_collection_Object_Shared_Pointer_Iterator->second->Set_Can_Commit(true);
241         }
242     }
243     if (!can_Commit) {
244         TX::test_counter += 1;
245         for (working_Map_collection_Object_Shared_Pointer_Iterator = working_Map_collection\&.begin(); working_Map_collection_Object_Shared_Pointer_Iterator != working_Map_collection\&.end(); working_Map_collection_Object_Shared_Pointer_Iterator++) {
246           
247             main_Process_Map_collection_Iterator  = TX::main_Process_Map_collection\&.find(working_Map_collection_Object_Shared_Pointer_Iterator->second->Get_Unique_ID());
248             (working_Map_collection_Object_Shared_Pointer_Iterator->second)->copy(working_Map_collection_Object_Shared_Pointer_Iterator->second, main_Process_Map_collection_Iterator->second);
249 
250         }
251         
252         _release_object_lock();
253 
254         return false;
255     } else {
256         /*
257          * Commit changes
258          */
259         for (working_Map_collection_Object_Shared_Pointer_Iterator = working_Map_collection\&.begin(); working_Map_collection_Object_Shared_Pointer_Iterator != working_Map_collection\&.end(); working_Map_collection_Object_Shared_Pointer_Iterator++) {
260             
261                 main_Process_Map_collection_Iterator = TX::main_Process_Map_collection\&.find((working_Map_collection_Object_Shared_Pointer_Iterator->second)->Get_Unique_ID());
262                 if (main_Process_Map_collection_Iterator != TX::main_Process_Map_collection\&.end()) {
263 
264                     (main_Process_Map_collection_Iterator->second)->copy(main_Process_Map_collection_Iterator->second, working_Map_collection_Object_Shared_Pointer_Iterator->second);
265                     main_Process_Map_collection_Iterator->second->increase_VersionNumber();
266 
267 
268                 } else {
269                     throw std::runtime_error(std::string("[RUNTIME ERROR : CAN'T FIND OBJECT COMMIT FUNCTION]"));
270 
271                 }
272         }
273 
274 
275         _release_object_lock();
276         this->th_exit();
277         return true;
278     }
279 }//Commit finish
.fi
.SS "std::map< int, int > TX::get_thread_Map ()\fC [private]\fP"

.PP
get_thread_Map returning and map to insert to the process_map_collection as an inner value get_thread_Map std::map, returning a map to store all unique ID from all objects from all transactions within the main process
.PP
\fBParameters:\fP
.RS 4
\fIthread_Map\fP std::map< int, int >, 
.RE
.PP

.PP
Definition at line 338 of file TX\&.cpp\&.
.PP
Referenced by _register()\&.
.PP
.nf
338                                       {
339     std::map< int, int > thread_Map;
340     return thread_Map;
341 }
.fi
.SS "int TX::getTest_counter ()"

.PP
getTest_counter TESTING ONLY!!! returning the value of the test_counter stored, number of rollbacks 
.PP
Definition at line 324 of file TX\&.cpp\&.
.PP
References test_counter\&.
.PP
.nf
324                         {
325     return TX::test_counter;
326 }
.fi
.SS "std::shared_ptr< \fBOSTM\fP > TX::load (std::shared_ptr< \fBOSTM\fP > object)"

.PP
load std::shared_ptr<OSTM>, returning an std::shared_ptr<OSTM> copy of the original pointer, to work with during transaction life time Register \fBOSTM\fP pointer into STM library
.PP
\fBParameters:\fP
.RS 4

Linux/O_STM/TX.cpp
/* 
 * File:   TX.cpp
 * Author: Zoltan Fuzesi
 * 
 * Created on December 18, 2017, 2:09 PM
 * TX cpp file methods implementations
 */
#include "TX.h"
#include <iostream>
/*!
 * \param static Global std::map main_Process_Map_collection store all transactional objects/pointers
 */
std::map<int, std::shared_ptr<OSTM> >TX::main_Process_Map_collection;
/*!
 * \param static Global std::map process_map_collection store all transactional objects/pointers
 */
std::map<int, std::map< int, int >> TX::process_map_collection;
/*!
 * \param static shared std:mutex register_Lock to protect writes into shared global collection
 */
std::mutex TX::register_Lock;
/*!
 * \param static Global counter for rollback
 */
int TX::test_counter = 0;
/*!
 * \brief Constructor
 * \param transaction_Number int, to store associated thread
 * \param _tx_nesting_level int, to store and indicate nesting level of transactions within transaction
 */
TX::TX(std::thread::id id) {
    this->transaction_Number = id;
    this->_tx_nesting_level = 0;
}
 /*!
  * \brief De-constructor
  */
TX::~TX() {
   
}
/*!
 * \brief Default copy constructor
 */
TX::TX(const TX& orig) {

}

/*!
 * \brief th_exit void, delete all std::shared_ptr<OSTM> elements from  working_Map_collection, that store pointers to working objects
 * \param working_Map_collection std::map, store std::shared_ptr<OSTM> transaction pointers
 */
void TX::th_exit() {

    if (this->_tx_nesting_level > 0) {
        /*
         * Active nested transactions running in background, do not delete anything yet
         */
    } else {
        /* 
         * Remove all elements map entries from transaction and clear the map
         */
        working_Map_collection.clear();
    }
}

/*!
 * \brief ostm_exit void, clear all elements from the shared global collections associated with the main process
 * \param main_Process_Map_collection std::map, store all std::shared_ptr<OSTM> from all transaction shared between multiple processes
 * \param process_map_collection std::map, store all unique id from all transaction within main process
 * DO NOT CALL THIS METHOD EXPLICITLY!!!!!! WILL DELETE ALL PROCESS  ASSOCIATED ELEMENTS!!!!
 */
void TX::ostm_exit() {
    std::map<int, std::shared_ptr<OSTM>>::iterator main_Process_Map_collection_Iterator;
     
    int ppid = getpid();
    std::map<int, std::map< int, int >>::iterator process_map_collection_Iterator = TX::process_map_collection.find(ppid);
    if (process_map_collection_Iterator != TX::process_map_collection.end()) {

        for (auto current = process_map_collection_Iterator->second.begin(); current != process_map_collection_Iterator->second.end(); ++current) {
            main_Process_Map_collection_Iterator = TX::main_Process_Map_collection.find(current->first);

            if (main_Process_Map_collection_Iterator != TX::main_Process_Map_collection.end()){
                /*
                 * Delete element from shared main_Process_Map_collection by object unique key value, shared_ptr will destroy automatically
                 */
                TX::main_Process_Map_collection.erase(main_Process_Map_collection_Iterator->first);      
            }
        }
        /*
         * Delete from Process_map_collection, Main process exits delete association with library
         */
        TX::process_map_collection.erase(process_map_collection_Iterator->first);
    }
}

/*!
 * \brief register void, receives an std::shared_ptr<OSTM> that point to the original memory space to protect from reca conditions
 * \param working_Map_collection std::map, store all the std::shared_ptr<OSTM> pointer in the transaction
 * \param main_Process_Map_collection std::map, store all std::shared_ptr<OSTM> from all transaction, used to lock and compare the objects
 * \param process_map_collection std::map, store all std::shared_ptr<OSTM> unique ID from all transaction, used to delete all pointers used by the main process, from all transaction before the program exit.
 * \param std::lock_guard use register_Lock(mutex) shared lock between all transaction
 * \param ppid int, store main process number 
 */
void TX::_register(std::shared_ptr<OSTM> object) {
    /*
     * MUST USE SHARED LOCK TO PROTECT SHARED GLOBAL MAP/COLLECTION 
     */
    std::lock_guard<std::mutex> guard(TX::register_Lock);
    
    /*
     * Check for null pointer !
     * Null pointer can cause segmentation fault!!!
     */
    if(object == nullptr){
        throw std::runtime_error(std::string("[RUNTIME ERROR : NULL POINTER IN REGISTER FUNCTION]") );
    }
    
    int ppid = getpid();
    std::map<int, std::map< int, int >>::iterator process_map_collection_Iterator = TX::process_map_collection.find(ppid);
    if (process_map_collection_Iterator == TX::process_map_collection.end()) {
        /*
         * Register main process/application to the global map
         */
        std::map< int, int >map =  get_thread_Map();
        TX::process_map_collection.insert({ppid, map});
        /*
         * Get the map if registered first time
         */
        process_map_collection_Iterator = TX::process_map_collection.find(ppid);
    }
    std::map<int, std::shared_ptr<OSTM>>::iterator main_Process_Map_collection_Iterator = TX::main_Process_Map_collection.find(object->Get_Unique_ID());
    if (main_Process_Map_collection_Iterator == TX::main_Process_Map_collection.end()) {
        /*
         * Insert to the GLOBAL MAP 
         */
        TX::main_Process_Map_collection.insert({object->Get_Unique_ID(), object});
        /*
         * Insert to the GLOBAL MAP as a helper to clean up at end of main process 
         */
        process_map_collection_Iterator->second.insert({object->Get_Unique_ID(), 1});
    } 


    std::map< int, std::shared_ptr<OSTM> >::iterator working_Map_collection_Object_Shared_Pointer_Iterator = working_Map_collection.find(object->Get_Unique_ID());
    if (working_Map_collection_Object_Shared_Pointer_Iterator == working_Map_collection.end()) {

        working_Map_collection.insert({object->Get_Unique_ID(), object->getBaseCopy(object)});
    }

}
/*!
 * \brief load std::shared_ptr<OSTM>, returning an std::shared_ptr<OSTM> copy of the original pointer, to work with during transaction life time
 * \param  working_Map_collection std::map, store all the std::shared_ptr<OSTM> pointer in the transaction
 */
std::shared_ptr<OSTM> TX::load(std::shared_ptr<OSTM> object) {

    std::map< int, std::shared_ptr<OSTM> >::iterator working_Map_collection_Object_Shared_Pointer_Iterator;
    /*
     * Check for null pointer !
     * Null pointer can cause segmentation fault!!!
     */
    if(object == nullptr){
        throw std::runtime_error(std::string("[RUNTIME ERROR : NULL POINTER IN LOAD FUNCTION]") );
    }

        working_Map_collection_Object_Shared_Pointer_Iterator = working_Map_collection.find(object->Get_Unique_ID());

    if (working_Map_collection_Object_Shared_Pointer_Iterator != working_Map_collection.end()) {

        return working_Map_collection_Object_Shared_Pointer_Iterator->second->getBaseCopy(working_Map_collection_Object_Shared_Pointer_Iterator->second);
        
    } else { throw std::runtime_error(std::string("[RUNTIME ERROR : NO OBJECT FOUND LOAD FUNCTION]") );}
}
/*!
 * \brief store void, receive an std::shared_ptr<OSTM> object to store the changes within the transaction, depends the user action
 * \param  working_Map_collection std::map, store all the std::shared_ptr<OSTM> pointer in the transaction
 */
void TX::store(std::shared_ptr<OSTM> object) {
    /*
     * Check for null pointer !
     * Null pointer can cause segmentation fault!!!
     */
    if(object == nullptr){
        throw std::runtime_error(std::string("[RUNTIME ERROR : NULL POINTER IN STORE FUNCTION]") );
    }
    
    std::map< int, std::shared_ptr<OSTM> >::iterator working_Map_collection_Object_Shared_Pointer_Iterator;

    working_Map_collection_Object_Shared_Pointer_Iterator = working_Map_collection.find(object->Get_Unique_ID());
    if (working_Map_collection_Object_Shared_Pointer_Iterator != working_Map_collection.end()) {

        working_Map_collection_Object_Shared_Pointer_Iterator->second = object;

    } else { std::cout << "[ERROR STORE]" << std::endl; }
}
/*!
 * \brief commit bool, returns boolean value TRUE/FALSE depends on the action taken within the function
 * \param  working_Map_collection std::map, store all the std::shared_ptr<OSTM> pointer in the transaction
 * \param main_Process_Map_collection std::map, store all std::shared_ptr<OSTM> from all transaction, used to lock and compare the objects
 * \param can_Commit bool, helps to make decision that the transaction can commit or rollback
 */
bool TX::commit() {

    bool can_Commit = true;
 
    /*
     * Dealing with nested transactions first 
     */
    if (this->_tx_nesting_level > 0) {
        _decrease_tx_nesting();
        return true;
    } 
    
    std::map< int, std::shared_ptr<OSTM> >::iterator working_Map_collection_Object_Shared_Pointer_Iterator;

    std::map<int, std::shared_ptr<OSTM>>::iterator main_Process_Map_collection_Iterator;
    for (working_Map_collection_Object_Shared_Pointer_Iterator = working_Map_collection.begin(); working_Map_collection_Object_Shared_Pointer_Iterator != working_Map_collection.end(); working_Map_collection_Object_Shared_Pointer_Iterator++) {

            main_Process_Map_collection_Iterator = TX::main_Process_Map_collection.find(working_Map_collection_Object_Shared_Pointer_Iterator->second->Get_Unique_ID());
            /*
             * Throws runtime error if object can not find
             */
            if(main_Process_Map_collection_Iterator == TX::main_Process_Map_collection.end())
            {
                throw std::runtime_error(std::string("[RUNTIME ERROR : CAN'T FIND OBJECT COMMIT FUNCTION]"));
            }

        /*
         * Busy wait WHILE object locked by other thread
         */
        while(!(main_Process_Map_collection_Iterator->second)->is_Locked());

        if (main_Process_Map_collection_Iterator->second->Get_Version() > working_Map_collection_Object_Shared_Pointer_Iterator->second->Get_Version()) {

            working_Map_collection_Object_Shared_Pointer_Iterator->second->Set_Can_Commit(false);
            can_Commit = false;
            break;
        } else {

            working_Map_collection_Object_Shared_Pointer_Iterator->second->Set_Can_Commit(true);
        }
    }
    if (!can_Commit) {
        TX::test_counter += 1;
        for (working_Map_collection_Object_Shared_Pointer_Iterator = working_Map_collection.begin(); working_Map_collection_Object_Shared_Pointer_Iterator != working_Map_collection.end(); working_Map_collection_Object_Shared_Pointer_Iterator++) {
          
            main_Process_Map_collection_Iterator  = TX::main_Process_Map_collection.find(working_Map_collection_Object_Shared_Pointer_Iterator->second->Get_Unique_ID());
            (working_Map_collection_Object_Shared_Pointer_Iterator->second)->copy(working_Map_collection_Object_Shared_Pointer_Iterator->second, main_Process_Map_collection_Iterator->second);

        }
        
        _release_object_lock();

        return false;
    } else {
        /*
         * Commit changes
         */
        for (working_Map_collection_Object_Shared_Pointer_Iterator = working_Map_collection.begin(); working_Map_collection_Object_Shared_Pointer_Iterator != working_Map_collection.end(); working_Map_collection_Object_Shared_Pointer_Iterator++) {
            
                main_Process_Map_collection_Iterator = TX::main_Process_Map_collection.find((working_Map_collection_Object_Shared_Pointer_Iterator->second)->Get_Unique_ID());
                if (main_Process_Map_collection_Iterator != TX::main_Process_Map_collection.end()) {

                    (main_Process_Map_collection_Iterator->second)->copy(main_Process_Map_collection_Iterator->second, working_Map_collection_Object_Shared_Pointer_Iterator->second);
                    main_Process_Map_collection_Iterator->second->increase_VersionNumber();


                } else {
                    throw std::runtime_error(std::string("[RUNTIME ERROR : CAN'T FIND OBJECT COMMIT FUNCTION]"));

                }
        }


        _release_object_lock();
        this->th_exit();
        return true;
    }
}//Commit finish

/*!
 * \brief _release_object_lock void, is get called from commit function, with the purpose to release the locks on all the objects participating in the transaction
 *  \param  working_Map_collection std::map, store all the std::shared_ptr<OSTM> pointer in the transaction
 *  \param main_Process_Map_collection std::map, store all std::shared_ptr<OSTM> from all transaction, used to release the lock on object
 */
void TX::_release_object_lock(){
    
    std::map< int, std::shared_ptr<OSTM> >::iterator working_Map_collection_Object_Shared_Pointer_Iterator;
    std::map<int, std::shared_ptr<OSTM>>::iterator main_Process_Map_collection_Iterator;
    for (working_Map_collection_Object_Shared_Pointer_Iterator = working_Map_collection.begin(); working_Map_collection_Object_Shared_Pointer_Iterator != working_Map_collection.end(); working_Map_collection_Object_Shared_Pointer_Iterator++) {

            main_Process_Map_collection_Iterator = TX::main_Process_Map_collection.find((working_Map_collection_Object_Shared_Pointer_Iterator->second)->Get_Unique_ID());
            if (main_Process_Map_collection_Iterator != TX::main_Process_Map_collection.end()) {
                /*
                 * Release object lock
                 */
                (main_Process_Map_collection_Iterator)->second->unlock_Mutex();
                
            } 
        }
}

/*!
 * \brief _increase_tx_nesting increase the value stored in _tx_nesting_level by one, indicate that the transaction nested
 * \param _tx_nesting_level int
 */
void TX::_increase_tx_nesting() {
      
    this->_tx_nesting_level += 1;
    // std::cout << "[this->_tx_nesting_level] = " << this->_tx_nesting_level << std::endl;
}
/*!
 * \brief _decrease_tx_nesting decrease the value stored in _tx_nesting_level by one, when outer transactions commiting
 * \param _tx_nesting_level int
 */
void TX::_decrease_tx_nesting() {
   // std::cout << "[this->_tx_nesting_level] = " << this->_tx_nesting_level << std::endl;
    this->_tx_nesting_level -= 1;
;
}
/*!
 * \brief getTest_counter TESTING ONLY!!! returning the value of the test_counter stored, number of rollbacks
 */
int TX::getTest_counter() {
    return TX::test_counter;
}
/*!
 * \brief _get_tx_number std::thread::id, returning the thread id that has assigned the given transaction
 * \param transaction_Number int
 */
const std::thread::id TX::_get_tx_number() const {
    return transaction_Number;
}
/*!
 * \brief get_thread_Map std::map, returning  a map to store all unique ID from all objects from all transactions within the main process
 * \param thread_Map std::map< int, int >, 
 */
std::map< int, int > TX::get_thread_Map() {
    std::map< int, int > thread_Map;
    return thread_Map;
}

/*!
 * ONLY FOR TESTING CHECK THE MAP AFTER THREAD EXIT AND ALL SHOULD BE DELETED!!!!!!!
 */
void TX::_print_all_tx() {

    std::cout << "[PRINTALLTHREAD]" << std::endl;
    std::map< int, std::shared_ptr<OSTM> >::iterator it;
    /*
     * All registered thread id in the TX global 
     */
    int ppid = getpid();
    std::map<int, std::map< int, int >>::iterator process_map_collection_Iterator = TX::process_map_collection.find(ppid);
    if (process_map_collection_Iterator != TX::process_map_collection.end()) {

        for (auto current = process_map_collection_Iterator->second.begin(); current != process_map_collection_Iterator->second.end(); ++current) {
            it = working_Map_collection.find(current->first);
            if(it != working_Map_collection.end()){
                std::cout << "[Unique number ] : " <<it->second->Get_Unique_ID() << std::endl;
            }

            
        }
     
    }
}